### ISSUE TITLE
Missing gradient when autograd called inside a function on Multi-GPU (eg gradient penalty)

### ISSUE THREAD

**Issue by tebesu** [pytorch/pytorch#16532]
## 🐛 Bug

Gradient is missing when calling torch.autograd.grad wrapped inside a function on multiple GPU's. (eg computing wgan gradient penalty). Calling torch.autograd.grad inline (not wrapped in a function) on multiple GPU's returns expected behavior.

## To Reproduce

Code below:
```python
import torch
import torch.nn as nn

torch.cuda.manual_seed_all(0)
torch.manual_seed(0)


def gradient_penalty(netD, x):
    """Functional Gradient Calculation"""
    output = netD(x)
    gradients = torch.autograd.grad(outputs=output, inputs=x,
                                    grad_outputs=x.new_ones(output.size()),
                                    create_graph=True, retain_graph=True)[0].mean()
    return gradients


net = nn.Linear(4, 1).cuda()
multigpu_net = nn.DataParallel(net, [0, 1])

x = torch.ones(2, 4, requires_grad=True).cuda()

print("Single GPU Functional")
net.zero_grad()
loss = gradient_penalty(net, x)
loss.backward()
print("Loss:", loss.item())
print("Grad:", [p.grad for p in net.parameters() if p.grad is not None])


print("\nMulti-GPU Functional")
multigpu_net.zero_grad()
loss = gradient_penalty(multigpu_net, x)
loss.backward()
print("Loss:", loss.item())
print("Grad:", [p.grad for p in net.parameters() if p.grad is not None])

print("\nMulti-GPU Inline")
multigpu_net.zero_grad()
output = multigpu_net(x)

# Compute grad inline
loss = torch.autograd.grad(outputs=output, inputs=x,
                            grad_outputs=x.new_ones(output.size()),
                            create_graph=True, retain_graph=True)[0].mean()
loss.backward()
print("Loss:", loss.item())
print("Grad:", [p.grad for p in net.parameters() if p.grad is not None])
```

<br>

Output for a single GPU calling `gradient_penalty` function
```
Single GPU Functional
Loss: -0.1287534236907959
Grad: [tensor([[0.2500, 0.2500, 0.2500, 0.2500]], device='cuda:0')]
```

<br>

MultiGPU calling `gradient_penalty` function
```
Multi-GPU Functional
Loss: -0.1287534236907959
Grad: [tensor([[0., 0., 0., 0.]], device='cuda:0')]
```

<br>

Multi-GPU calling `autograd.grad` inline (not inside a function)
```
Multi-GPU Inline
Loss: -0.1287534236907959
Grad: [tensor([[0.2500, 0.2500, 0.2500, 0.2500]], device='cuda:0'), tensor([0.], device='cuda:0')]
```

<br>

## Expected behavior

The gradient should be accumulated when calling `autograd.grad` from inside another function. All outputs gradients from the script should be the same.

## Environment

 - PyTorch Version (e.g., 1.0): 1.0.9
 - OS (e.g., Linux): Ubuntu 16.04.3 LTS
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Python version: 3.6
 - CUDA/cuDNN version: 9.0/6
 - GPU models and configuration:
GPU 0: GeForce GTX 1080
GPU 1: GeForce GTX 1080
GPU 2: GeForce GTX 1080
GPU 3: GeForce GTX 1080

**Comment by mrshenli** [comment]
It is because, when exiting the function, the `output` will be deleted, which triggers [recursive graph function deletion](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/function.cpp#L77) and deletes the `next_edges_` from `BroadcastBackward` to `AccumulateGrad`.  As a result,  the gradient will not be accumulated into `net`. You can avoid this by returning both the `gradient` and the `output`. Let me see if we can/should add reference to the `output`.

@gchanan Question: do we need to consider the case where the [reference count](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/function.cpp#L51) is larger than 1 when deleting the function? I do see the reference count equals to 2 when running the `Compute grad inline` part in the above example. But maybe it will disappear after fixing the reference problem.

**Comment by gchanan** [comment]
curious: does this work pre https://github.com/pytorch/pytorch/commit/89d56ae435373e54932566806c8432634e6a5cfa
?

**Comment by tebesu** [comment]
@gchanan 
This problem does not occur in Pytorch 0.4. I have not tried the specific commit you mentioned.

**Comment by gchanan** [comment]
cc @resistor.

**Comment by mrshenli** [comment]
It seems to be sth wrong with the reference count when used together with data parallel. Initially, there are two `TBackward` replicas and two `ExpandBackward` replicas point to `BroadcastBackward`. So, the reference count for `BroadcastBackward` is 4, but later (don't know where yet), the reference count drops to 2.  Deleting the two `ExpandBackward` triggers [`gatherFunctions()`](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/function.cpp#L50) to destroy the edges from `BroadcastBackward` to `AccumulateGrad`, while there are still two `TBackward` point to `BroadcastBackward`.

**Comment by mrshenli** [comment]
After some digging, I think the cause of this problem is that the `use_count` of function `shared_ptr` are incorrect on `C++` side as we [create multiple `shared_ptr`](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/python_variable.cpp#L57) from raw pointers when `replicate` modules. It was not a problem pre 89d56ae because python knows it correctly, as we [increased the ref count](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/python_function.cpp#L1095) on python side. So, when used together with data parallel, we cannot use the [`use_count` on `shared_ptr`](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/function.cpp#L49) to determine when an edge can be deleted.

**Comment by pietern** [comment]
There are alternative repro scripts in #19024 (contributed by @hukkelas and @thuyen).

**Comment by ezyang** [comment]
Fixed by https://github.com/pytorch/pytorch/pull/22983
